osu.Game/Database/RealmFileStore.cs

```diff
@@ -50,20 +51,24 @@ public RealmFile Add(Stream data, Realm realm, bool addToRealm = true, bool pref
             var file = existing ?? new RealmFile { Hash = hash };
 
             if (!checkFileExistsAndMatchesHash(file))
-                copyToStore(file, data, preferHardLinks);
+                copyToStore(file, data, preferHardLinks, preferCopyOnWrite);
 
             if (addToRealm && !file.IsManaged)
                 realm.Add(file);
 
             return file;
         }
 
-        private void copyToStore(RealmFile file, Stream data, bool preferHardLinks)
+        private void copyToStore(RealmFile file, Stream data, bool preferHardLinks, bool preferCopyOnWrite)
         {
-            if (data is FileStream fs && preferHardLinks)
+            if (data is FileStream fs)
             {
+                // Prefer CoW over hard link because it provides better isolation.
+                if (preferCopyOnWrite && CopyOnWriteHelper.TryCloneFile(Storage.GetFullPath(file.GetStoragePath(), true), fs.Name))
+                    return;
+
                 // attempt to do a fast hard link rather than copy.
-                if (HardLinkHelper.TryCreateHardLink(Storage.GetFullPath(file.GetStoragePath(), true), fs.Name))
+                if (preferHardLinks && HardLinkHelper.TryCreateHardLink(Storage.GetFullPath(file.GetStoragePath(), true), fs.Name))
                     return;
```

Comment:
As it currently stands, `TryCloneFile` will always succeed on Win11 24H2+, preventing the hard link fallback. Maybe `preferHardLinks` and `preferCopyOnWrite` could be set iff CoW/hardlinks are actually available, instead of always being `true`. The functionality (`CheckAvailability`) is already there, but it's only used for displaying messages to the user.

You can't really modify `TryCloneFile` to check CoW availability on every file, as that would probably kill performance.

Reply:
I'm unsure about extending to other platforms. On older Windows it requires complicated `DeviceIoControl` syscall which is not interesting. MacOS provides a separated `clonefile` syscall. For Linux I found some information about the `cp` command defaults to CoW on btrfs, but no general information for the underlying syscalls.

> preventing the hard link fallback.

Speaking for the Windows implementation: if real CoW action is not available under the hood, then hard linking will probably fail too because of the requirements. In this situation, the `CopyFile` syscall will just "do the right thing": fallback to regular copying. CoW availability don't need to be distinguished at all, if `CopyOnWriteHelper.CheckAvailability` decides `CopyFile` is doing the best thing.

For other implementations when cloning requires another syscall, it's failure need to be handled.

Reply:
You missed my point. `TryCloneFile` will always succeed, even if the file is not cloned (win11 24H2+), so `TryCreateHardLink` will never be called. As it stands, hard links will never be created on win11 24H2. Not on ReFS, not on NTFS, not on DevDrive. This is a regression.

Reply:
Ah I see. `PreferCopyOnWrite`/`PreferHardLinks` are always passed as true, disconnected from the result of `CheckAvailability`. `CheckAvailability` is used for display only then.